dNet/ClientPackets.cpp

```diff
@@ -5,119 +5,136 @@
+		reputationLdf->WriteToPacket(data);
+
+		//Compress the data before sending:
+		const uint32_t reservedSize = ZCompression::GetMaxCompressedLength(data.GetNumberOfBytesUsed());
+		uint8_t* compressedData = new uint8_t[reservedSize];
+
+		if (!compressedData) {
+			throw std::runtime_error("Failed to allocate memory for compressed data");
 		}
-	}
 
-	bool hasRemoteInputInfo{};
-	if (inStream.Read(hasRemoteInputInfo) && hasRemoteInputInfo) {
-		inStream.Read(update.remoteInputInfo.m_RemoteInputX);
-		inStream.Read(update.remoteInputInfo.m_RemoteInputY);
-		inStream.Read(update.remoteInputInfo.m_IsPowersliding);
-		inStream.Read(update.remoteInputInfo.m_IsModified);
-	}
+		size_t size = ZCompression::Compress(data.GetData(), data.GetNumberOfBytesUsed(), compressedData, reservedSize);
 
-	return update;
-}
+		assert(size <= reservedSize);
 
-ChatModerationRequest ClientPackets::HandleChatModerationRequest(Packet* packet) {
-	CINSTREAM_SKIP_HEADER;
-	
-	ChatModerationRequest request;
+		bitStream.Write<uint32_t>(size + 9); //size of data + header bytes (8)
+		bitStream.Write<uint8_t>(1);         //compressed boolean, true
+		bitStream.Write<uint32_t>(data.GetNumberOfBytesUsed());
+		bitStream.Write<uint32_t>(size);
 
-	inStream.Read(request.chatLevel);
-	inStream.Read(request.requestID);
+		/**
+		 * In practice, this warning serves no purpose for us.  We allocate the max memory needed on the heap
+		 * and then compress the data.  In the off chance that the compression actually increases the size,
+		 * an assertion is done to prevent bad data from being saved or sent.
+		 */
+	#pragma warning(disable:6385) // C6385 Reading invalid data from 'compressedData'.
+		bitStream.WriteAlignedBytes(compressedData, size);
+	#pragma warning(default:6385)
+		delete[] compressedData;
+	};
 
-	for (uint32_t i = 0; i < 42; ++i) {
-		uint16_t character;
-		inStream.Read(character);
-		request.receiver.push_back(static_cast<uint8_t>(character));
-	}
+	void ChatModerationString::Serialize(RakNet::BitStream &bitStream) const {
+		bitStream.Write<uint8_t>(rejectedWords.empty()); // Accepted
+		bitStream.Write<uint16_t>(0); // padding
+
+		bitStream.Write(chatChannel);
+		bitStream.Write(chatMode);
+
+		bitStream.Write(LUWString(receiver, 42));
 
-	if (!request.receiver.empty()) {
-		if (std::string(request.receiver.c_str(), 4) == "[GM]") { // Shift the string forward if we are speaking to a GM as the client appends "[GM]" if they are
-			request.receiver = std::string(request.receiver.c_str() + 4, request.receiver.size() - 4);
+		for (auto it : rejectedWords) {
+			bitStream.Write<uint8_t>(it.first); // start index
+			bitStream.Write<uint8_t>(it.second); // length
 		}
+
+		// Pad out the rest of the packet
+		// The client expects 64 items, so we need to write 64 - rejectedWords.size() empty items
+		for (int i = rejectedWords.size(); 64 > i; i++) {
+			bitStream.Write<uint16_t>(0);
+		}
```

Comment:
The first loop should make sure only 64 items are being written in this case, writing more will not be read or acknowledged.  If we're going to be making some minor logic changes let's get a few of them that are minor while we're here